src/main/java/nl/knaw/dans/ingest/core/service/ManifestHelper.java

```diff
@@ -0,0 +1,102 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static gov.loc.repository.bagit.hash.StandardSupportedAlgorithms.SHA1;
+
+public class ManifestHelper {
+
+    static public void addSha1File(Bag bag) throws NoSuchAlgorithmException, IOException {
+        var manifests = bag.getPayLoadManifests();
+        var algorithms = manifests.stream().map(Manifest::getAlgorithm);
+        if (algorithms.anyMatch(SHA1::equals))
+            return;
+
+        var payloadFilesMap = Hasher.createManifestToMessageDigestMap(List.of(SHA1));
+        var payloadVisitor = new CreatePayloadManifestsVistor(payloadFilesMap, true);
+        Files.walkFileTree(PathUtils.getDataDir(bag), payloadVisitor);
+        manifests.addAll(payloadFilesMap.keySet());
+        ManifestWriter.writePayloadManifests(manifests, PathUtils.getBagitDir(bag), bag.getRootDir(), bag.getFileEncoding());
+
+        updateTagManifests(bag);
+    }
+
+    private static void updateTagManifests(Bag bag) throws NoSuchAlgorithmException, IOException {
+        var algorithms = bag.getTagManifests().stream()
+            .map(Manifest::getAlgorithm)
+            .collect(Collectors.toList());
+        var tagFilesMap = Hasher.createManifestToMessageDigestMap(algorithms);
+        var bagRootDir = bag.getRootDir();
+        var tagVisitor = new CreateTagManifestsVistor(tagFilesMap, true) {
+            @Override
+            public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+                /*
+                 * Fix for EASY-1306: a tag manifest must not contain an entry for itself, as this is practically
+                 * impossible to calculate. It could in theory contain entries for other tag manifests. However,
+                 * the CreateTagManifestsVistor, once it finds an entry for a tag file in ONE of the tag manifests,
+                 * will add an entry in ALL tag manifests.
+                 *
+                 * Therefore we adopt the strategy NOT to calculate any checksums for the tag manifests themselves.
+                 *
+                 * Update: this is actually required in V1.0: https://tools.ietf.org/html/rfc8493#section-2.2.1
+                 */
+                var isTagManifest = bagRootDir.relativize(path).getNameCount() == 1 &&
+                    path.getFileName().toString().startsWith("tagmanifest-");
+                if (isTagManifest) return FileVisitResult.CONTINUE;
+                else return super.visitFile(path, attrs);
+            }
+        };
+        Files.walkFileTree(bagRootDir, tagVisitor);
+        bag.getTagManifests().clear();
+        bag.getTagManifests().addAll(tagFilesMap.keySet());
+        ManifestWriter.writeTagManifests(bag.getTagManifests(), PathUtils.getBagitDir(bag), bagRootDir, bag.getFileEncoding());
+    }
+
+    static public Map<Path, String> getFilePathToSha1(Deposit deposit) {
```

Comment:
This method is moved here from `FileElement` as is. I'm hesitating to change the argument to a bag.

Reply:
Deposit is used to find the bag directory in line 97, if you can do that without the Deposit object, I guess you could refactor this method.